### DISCUSSION TITLE
Wasm support

### DISCUSSION THREAD

**Discussion by trebonian** [bottlerocket-os/bottlerocket#2609] [Ideas]
Are there any plans to have bottlerocket support Cloud Native Webassembly using container runtimes such as "wasmedge"

**Comment by jpmcb** [comment]
Hi @trebonian - thanks for dropping this suggestion!!

Can you provide abit more of what you're looking for here? Are you hoping to use something like `wasmedge` as a replacement runtime in kubernetes?

The `wasmtime` project already supports running web assembly apps in kubernetes ontop of existing container technology, [including `containerd`](https://wasmedge.org/book/en/use_cases/kubernetes/cri/containerd.html) _but_ it requires `crun` to be configured in containred as the _underlying_ container-runtime. Currently, `containerd` is used as the container-runtime. In essence, looks like the abstraction would become

```
crun -> conatinerd -> wasi-container -> wasm-application
```

You may be interested in https://github.com/bottlerocket-os/bottlerocket/issues/835 where we propose extending the bottlerocket to support different runtimes. Until then, I don't believe this will be possible without building a different variant that compiles and packages `crun` and configures `containerd` to use it.

**Comment by arnaldo2792** [comment]
Adding to what @jpmcb , you should keep an eye on https://github.com/bottlerocket-os/bottlerocket/issues/2669, this will help developers create their own variants and change / use whatever runtimes they want :+1:

**Comment by trebonian** [comment]
I was hoping to have a build parameter to include both runtimes in the build, and then an runtime setting to select one or the other.

Is it possible to run both runtimes on the same server as kubernetes will need to schedule the workloads to each independently...

**Comment by jpmcb** [comment]
> Is it possible to run both runtimes on the same server as kubernetes will need to schedule the workloads to each independently...

Yes, it is possible to have multiple runtimes on a cluster and typically is [implemented through `RuntimeClasses`](https://kubernetes.io/docs/concepts/containers/runtime-class/) where workloads can choose the runtime they need. But in order to achieve a balance of security and usability, the main bottlerocket variants only support & ship the `containerd` runtime.

Like Arnaldo said, in the future, when out-of-tree variants of bottlerocket are possible, it will be much easier to change and build the runtime that's in use.
